packages/edit-site/src/components/url-query-controller/index.js

```diff
@@ -3,74 +3,63 @@
  */
 import { useEffect } from '@wordpress/element';
 import { useDispatch, useSelect } from '@wordpress/data';
-import { getQueryArg, addQueryArgs, removeQueryArgs } from '@wordpress/url';
 
 /**
  * Internal dependencies
  */
+import { useLocation, useHistory } from '../routes';
 import { store as editSiteStore } from '../../store';
 
 export default function URLQueryController() {
 	const { setTemplate, setTemplatePart, showHomepage, setPage } = useDispatch(
 		editSiteStore
 	);
+	const history = useHistory();
+	const {
+		params: { postId, postType },
+	} = useLocation();
+	const { getPage, getEditedPostId, getEditedPostType } = useSelect(
+		editSiteStore
+	);
 
-	// Set correct entity on load.
+	// Set correct entity on page navigation.
 	useEffect( () => {
-		const url = window.location.href;
-		const postId = getQueryArg( url, 'postId' );
-
-		if ( ! postId ) {
-			showHomepage();
-			return;
-		}
+		let isMounted = true;
 
-		const postType = getQueryArg( url, 'postType' );
 		if ( 'page' === postType || 'post' === postType ) {
 			setPage( { context: { postType, postId } } ); // Resolves correct template based on ID.
 		} else if ( 'wp_template' === postType ) {
 			setTemplate( postId );
 		} else if ( 'wp_template_part' === postType ) {
 			setTemplatePart( postId );
 		} else {
-			showHomepage();
-		}
-	}, [] );
-
-	// Update page URL when context changes.
-	const pageContext = useCurrentPageContext();
-	useEffect( () => {
-		const newUrl = pageContext
-			? addQueryArgs( window.location.href, pageContext )
-			: removeQueryArgs( window.location.href, 'postType', 'postId' );
+			showHomepage().then( () => {
+				if ( ! isMounted ) {
+					return;
+				}
 
-		window.history.replaceState( {}, '', newUrl );
-	}, [ pageContext ] );
+				const page = getPage();
+				const editedPostId = getEditedPostId();
+				const editedPostType = getEditedPostType();
 
-	return null;
-}
-
-function useCurrentPageContext() {
-	return useSelect( ( select ) => {
-		const { getEditedPostType, getEditedPostId, getPage } = select(
-			editSiteStore
-		);
-
-		const page = getPage();
-		let _postId = getEditedPostId(),
-			_postType = getEditedPostType();
-		// This doesn't seem right to me,
-		// we shouldn't be using the "page" and the "template" in the same way.
-		// This need to be investigated.
-		if ( page?.context?.postId && page?.context?.postType ) {
-			_postId = page.context.postId;
-			_postType = page.context.postType;
+				if ( page?.context?.postId && page?.context?.postType ) {
+					history.replace( {
+						postId: page.context.postId,
+						postType: page.context.postType,
+					} );
+				} else if ( editedPostId && editedPostType ) {
+					history.replace( {
+						postId: editedPostId,
+						postType: editedPostType,
+					} );
+				}
+			} );
```

Comment:
So this is where there's a client side redirection? (https://github.com/WordPress/gutenberg/issues/36873). I understand this is due to be replaced, is that going to be in 5.9?

I think this is a bit of tech debt because of how in all other cases the store state is reactive, but here it gets set first and the call to `history.replace` happens after. As long as it works right now, it's ok to leave it as is in this PR.

If we can't solve #36873, there might be a few things that can be done to tidy it up. I was wondering if it could be a selector like `getHomepage`, and then we handle routing based on what that returns. The state could then update as a result of that routing. Also feels like it could be done before the React App mounts to avoid any other effects triggering.

As a side note, the request to `/wp/v2/settings` that happens from `showHomepage` is possibly not using the preloaded data. I'm still seeing a network request in my dev tools. It probably wasn't an issue introduced here though, so let's follow up on that.

Reply:
> I understand this is due to be replaced, is that going to be in 5.9?

I don't think so, unless someone is willing to work on that.

> I was wondering if it could be a selector like `getHomepage`, and then we handle routing based on what that returns. The state could then update as a result of that routing. Also feels like it could be done before the React App mounts to avoid any other effects triggering.

Sounds interesting! Would you mind try implementing this? Either push directly to this PR or to a different PR.

Reply:
I'm not sure of all the permutations, but I'll give it a go 😄